File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.js
function solution(land) {
    class Queue{
    constructor(){
        this.q = [];
        this.head = -1;
        this.tail = -1;
    }
    empty(){
        return this.head === this.tail;
    }
    size(){
        return this.tail - this.head;
    }
    push(item){
        this.q.push(item);
        this.tail++;
    }
    pop(){
        if(this.empty()) console.log("empty");
        return this.q[++this.head];
    }
}
    var answer = [];
    
    let n = land.length;
    let m = land[0].length;
    let visited = Array.from(Array(n) , () => new Array(m).fill(0));
    let dp = new Array(m).fill(0);
    let queue = new Queue();
    
    let dx = [1 , -1 , 0 , 0];
    let dy = [0 , 0 , 1 , -1];
    
    function inRange(x , y){
        return 0 <= x && x < n && 0 <= y && y < m;
    }
    
    function canGo(x , y){
        return inRange(x , y) && !visited[x][y] && land[x][y] === 1;
    }
    
    function bfs(row , col , visited){
        let count = 1;
        queue.push([row , col]);
        visited[row][col] = true;
        let vi = new Set();
        
        while(queue.size() > 0){
            let [a , b] = queue.pop();
            vi.add(b);
            for(let i = 0; i < 4; i++){
                let nx = a + dx[i];
                let ny = b + dy[i];
                if(canGo(nx,  ny)){
                    vi.add(ny);
                    queue.push([nx , ny]);
                    visited[nx][ny] = true;
                    count++;
                }
            }
        }
        for(let v of vi){
            dp[v] += count;
        }
        
        return count;
    }
    
    function move(col , visited){
        let num = 0;
        for(let row = 0; row < n; row++){
            if(land[row][col] === 1 && !visited[row][col]){
                let t = bfs(row , col , visited);
            }
            else continue;
        }
    }
    
    for(let i = 0; i < m; i++){
        move(i , visited);
    }
    
    return Math.max(...dp);
}

File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.py
from collections import deque


def solution(land):
    answer = 0
    n, m = len(land), len(land[0])
    
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    
    # 각 열의 기름 총량
    result = [0 for i in range(m)]
    # 방문 기록
    visited = [[0 for _ in range(m)] for _ in range(n)]
    
    def bfs(pos):
        a, b = pos
        cnt = 1
        visited[a][b] = 1

        q = deque()
        q.append(pos)
        
        # 석유가 존재하는 열 저장 (중복 방지)
        oil_covered = set()
        oil_covered.add(b)
        
        while q:
            x, y = q.popleft()
            for i in range(4):
                nx, ny = x + dx[i], y + dy[i]
                if 0 <= nx < n and 0 <= ny < m and visited[nx][ny] == 0 and land[nx][ny] == 1:
                    q.append((nx, ny))
                    visited[nx][ny] = 1
                    cnt += 1
                    oil_covered.add(ny)
                    
        for c in oil_covered:
            result[c] += cnt
            
    for i in range(m):
        for j in range(n):
            if land[j][i] == 1 and not visited[j][i]:
                bfs((j, i))
    
    answer = max(result)
    
    return answer